### PR TITLE
Solves window related status bar & rotation issues

### PR DESCRIPTION
**Pull request by trooper2013**
@bhariharan based on our offline discussions about issues related to multiple windows, I'm finally  changing the window manager implementation to move away from the current strategy of adjusting transparency of windows. This fix involves moving away from the current approach of adjusting transparency of windows, to a model where the windows will show and hide instead.   Side effect  - window transitions may not be as smooth.

### CODE REVIEW

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SalesforceSDKManager.m [hunk 1]
```diff
@@ -830,11 +833,18 @@ - (void)dismissSnapshot
             self.snapshotDismissalAction(_snapshotViewController);
             if ([SFSecurityLockout isPasscodeNeeded]) {
                 [SFSecurityLockout validateTimer];
+            } else{
+                [SFSecurityLockout validateTimer];
+                [lastActiveWindow presentWindow];
             }
         } else {
-            [[SFSDKWindowManager sharedManager].snapshotWindow dismissWindowAnimated:NO withCompletion:^{
+            __block SFSDKWindowContainer *activeWindow = lastActiveWindow;
```
Comment (trooper2013) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SalesforceSDKManager.m]: Snapshot window needs special handling. We could be coming back to a main window or a passcode window. lastActiveWindow is captured before snapshot is presented.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFSDKOAuthClient.m [hunk 2]
```diff
@@ -55,7 +55,26 @@
 static NSString * const kSFRevokePath = @"/services/oauth2/revoke";
 
 static Class<SFSDKOAuthClientProvider> _clientProvider = nil;
+@interface UINavigationController(SFSDKOAuth)
+@end
+
+@implementation UINavigationController(SFSDKOAuth)
```
Comment (trooper2013) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFSDKOAuthClient.m]: allow us to delegate the call to the  login view controller subclass if required.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Views/SFSDKWindowManager.m [hunk 3]
```diff
@@ -36,9 +36,7 @@ @interface SFSDKWindowManager()<SFSDKWindowContainerDelegate>
 
 @property (nonatomic, strong) NSHashTable *delegates;
 @property (nonatomic, strong,readonly) NSMapTable<NSString *,SFSDKWindowContainer *> * _Nonnull namedWindows;
-
-- (void)makeTransparentWithCompletion:(SFSDKWindowContainer *)window completion:(void (^)(void))completion;
-- (void)makeOpaqueWithCompletion:(SFSDKWindowContainer *)window completion:(void (^)(void))completion;
+@property (nonatomic, strong,readonly) NSMapTable<UIWindow *,SFSDKWindowContainer *> * _Nonnull reverseLookupTable;
```
Comment (trooper2013) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Views/SFSDKWindowManager.m]: Maintains a reverse map to test for managed windows throughout the window management.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Views/SFSDKWindowContainer.h [hunk 4]
```diff
@@ -93,15 +93,27 @@ typedef NS_ENUM(NSInteger, SFSDKWindowType) {
...
  * Make window visible
  */
-- (void)dismissWindowAnimated:(BOOL)animated withCompletion:(void (^_Nullable)(void))completion;
+- (void)dismissWindowWithCompletion:(void (^_Nullable)(void))completion;
+
+
+/**
+ * Make window visible, set alpha to 1.0 invoke completion block
+ */
+- (void)presentWindowAnimated:(BOOL)animated withCompletion:(void (^_Nullable)(void))completion SFSDK_DEPRECATED(6.0, 7.0, "Use presentWindowWithCompletion instead.");
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Views/SFSDKWindowContainer.h]: First parameter of `SFSDK_DEPRECATED` should be `6.2`.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Views/SFSDKWindowContainer.h [hunk 5]
```diff
@@ -93,15 +93,27 @@ typedef NS_ENUM(NSInteger, SFSDKWindowType) {
...
+
+/**
+ * Make window visible
+ */
+- (void)dismissWindowAnimated:(BOOL)animated withCompletion:(void (^_Nullable)(void))completion SFSDK_DEPRECATED(6.0, 7.0, "Use dismissWindowWithCompletion instead.");
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Views/SFSDKWindowContainer.h]: First parameter of `SFSDK_DEPRECATED` should be `6.2`.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Views/SFSDKWindowManager.m [hunk 6]
```diff
@@ -91,8 +83,10 @@ - (void)setMainUIWindow:(UIWindow *) window {
     SFSDKWindowContainer *container = [[SFSDKWindowContainer alloc] initWithWindow:window name:kSFMainWindowKey];
     container.windowType = SFSDKWindowTypeMain;
     container.windowDelegate = self;
-    container.window.alpha = 1.0;
+    container.viewController = [[SFSDKRootController alloc] init];
+    [container.window setHidden:NO];
```
Comment (khawkins) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Views/SFSDKWindowManager.m]: Minor: use property syntax here?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Views/SFSDKWindowManager.m [hunk 7]
```diff
@@ -289,55 +263,66 @@ - (SFSDKWindowContainer *)createPasscodeWindow {
...
+    [self.reverseLookupTable setObject:container forKey:container.window];
     return container;
 }
 
 -(UIWindow *)createDefaultUIWindow {
     UIWindow *window = [[UIWindow alloc] initWithFrame:UIScreen.mainScreen.bounds];
-    [window setAlpha:0.0];
+    window.hidden = YES;
+    window.backgroundColor = [UIColor whiteColor];
+    //[window.window resignKeyWindow];
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Views/SFSDKWindowManager.m]: Remove commented out code.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController.m [hunk 8]
```diff
@@ -236,9 +239,21 @@ - (UIView *)createTitleItem {
     return item;
 }
 
+- (BOOL)shouldAutorotate {
+    return YES;
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController.m]: Why did we change this?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController.m [hunk 9]
```diff
@@ -236,9 +239,21 @@ - (UIView *)createTitleItem {
...
+}
+
+- (UIInterfaceOrientationMask)supportedInterfaceOrientations {
+    return UIInterfaceOrientationMaskAll;
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController.m]: Why did we change this? I'd prefer if we locked down to portrait mode as the default within the SDK.

**Pull request merged**